### DISCUSSION TITLE
error: could not connect to database

### DISCUSSION THREAD

**Discussion by Guneetsinghtuli** [osm-search/Nominatim#3007] [Q&A]
I am currently using ubuntu 20 in a virtual server since setting up on Mac was becoming difficult, While importing the osm file to nominatim which I downloaded from [GeoFabrik](https://download.geofabrik.de/). I am facing this error 

`root@ubuntu-s-1vcpu-1gb-blr1-01:~/nominatim-planet# nominatim import --osm-file bahamas-latest.osm.pbf 2>&1 | tee setup.log
2023-03-11 11:41:19: Using project directory: /root/nominatim-planet
2023-03-11 11:41:22: Creating database
createdb: error: could not connect to database template1: FATAL:  role "root" does not exist
2023-03-11 11:41:22: FATAL: Creating new database failed.`

I doubled check whether I have created the user from [here](https://nominatim.org/release-docs/develop/appendix/Install-on-Ubuntu-22/#setting-up-postgresql).

On running the following command it is showing me a role already exists but it is also showing me some unpexted things
`root@ubuntu-s-1vcpu-1gb-blr1-01:~/nominatim-planet# sudo -u postgres createuser -s $USERNAME
could not change directory to "/root/nominatim-planet": Permission denied
createuser: error: creation of new role failed: ERROR:  role "nominatim" already exists`

Can someone help me out with this?

**Comment by mtmail** [comment]
The [instructions](https://nominatim.org/release-docs/develop/appendix/Install-on-Ubuntu-22/#creating-dedicated-user-accounts) include a warning `Never, ever run the installation as a root user. You have been warned.` 

Hopefully you just need to switch to your nominatim user account `sudo su nominatim` to run the script. I'd move or recreate the `nominatim-planet` project directory in user nominatim's home directory to avoid file permission errors.

**Comment by Guneetsinghtuli** [comment]
`nominatim@ubuntu-s-1vcpu-1gb-blr1-01:~/nominatim-planet$ sudo nominatim import --osm-file /srv/nominatim/nominatim-planet/bahamas-latest.osm.pbf 2>&1 | tee setup.log
Using project directory: /srv/nominatim/nominatim-planet
**Creating database createdb: error: could not connect to database template1: FATAL:  role "root" does not exist**
FATAL: Creating new database failed.
**nominatim@ubuntu-s-1vcpu-1gb-blr1-01**:~/nominatim-planet$ pwd
/srv/nominatim/nominatim-planet
nominatim@ubuntu-s-1vcpu-1gb-blr1-01:~/nominatim-planet$ `

I tried it with the nominatim user but still the same issue.

### Suggestion
And while following the [documentation](https://nominatim.org/release-docs/latest/appendix/Install-on-Ubuntu-20/) I think we can include the following steps of 
1. **Creating password** for user nominatim 
2. **Giving nominatim user sudo access** as by default it cannot run sudo commands 
3. **Switching the user** and few commands after that also might need change as they will also need sudo access or we should run them as sudo user

If you think this can be helpful would love to make a contribution to it

**Reply by mtmail** [reply]
Running `sudo nominatim import ...` will essentially run as `root`. The nominatim user doesn't need sudo power.

```
$ lsb_release -a
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.1 LTS
Release:	22.04
Codename:	jammy

$ whoami
nominatim

$ sudo whoami
root
```

You've created a `nominatim` user and can login as `nominatim`, that's a good step forward.

Try running the import command with `sudo`. If the (postgresql) error is `role "nominatim" does not exist` or similar then you can create a `nominatim` user in postgresql first.

```
sudo -u postgres createuser -s nominatim
sudo -u postgres createuser www-data
```

The `-s` gives the postgresql user 'superuser' power (https://www.postgresql.org/docs/14/app-createuser.html), that's similar to sudo in Ubuntu: the user can now create and delete databases among other powers. It's confusing that Postgresql calls their comment `createuser` and not something like `pg_createuser`, and can be confusing that postgresql has yet another user- and permission-system separate from the operating system.
